src/renderer_common/render_markers_symbolizer.cpp

```diff
@@ -130,18 +129,69 @@ struct render_marker_symbolizer_visitor
 
         boost::optional<svg_path_ptr> const& stock_vector_marker = mark.get_data();
         svg_path_ptr marker_ptr = *stock_vector_marker;
-        bool is_ellipse = false;
 
-        svg_attribute_type s_attributes;
-        auto const& r_attributes = get_marker_attributes(*stock_vector_marker, s_attributes);
+        std::shared_ptr<svg_attribute_type> r_attributes;
+
+        // Look up the feature/symbolizer attributes from the cache.
+        // We are using raw symbolizer pointer as a cache key. This is a hack but
+        // works in practice due to symbolizers being created all at once during style loading.
+        markers_symbolizer const* attr_key = &sym_;
+
+        auto attr_it = cached_attributes_.find(attr_key);
+        if (attr_it == cached_attributes_.end())
+        {
+            svg_attribute_type s_attributes;
+            r_attributes = std::make_shared<svg_attribute_type>(get_marker_attributes(*stock_vector_marker, s_attributes));
+
+            // We can only cache the attributes using the given key if no expressions are used in properties.
+            // Otherwise the expressions may refer to feature-specific values.
+            bool cacheable = std::all_of(
+                sym_.properties.begin(), sym_.properties.end(),
+                [](symbolizer_base::cont_type::value_type const& key_prop) { return !is_expression(key_prop.second); }
+            );
+            if (cacheable)
+            {
+                if (cached_attributes_.size() > cache_size)
+                {
+                    cached_attributes_.erase(cached_attributes_.begin());
+                }
+                attr_it = cached_attributes_.emplace(attr_key, r_attributes).first;
+            }
+        }
+        else
+        {
```

Comment:
Personal taste, but I'd rather have the smaller conditions first. Since this is just style do what you feel best.